### ISSUE TITLE
Allow generic type constraints

### ISSUE THREAD

**Issue by rkone** [dotnet/aspnetcore#8433]
We should be able to add generic constraints to our typeparam. For example:
`@typeparam TEntity where TEntity : IEntity`

As of preview3 gives the error "Unexpected literal following the 'typeparam' directive".

**Comment by stsrki** [comment]
I also think this is a much needed feature.

Any comment on this?

**Comment by prosser** [comment]
Without this feature, developers will be creating a lot of boilerplate components to workaround C# 8.0 #nullable limitations.

for example, take this overly-simplified .razor:

```
@typeparam TEnum
@typeparam TContainer
@inherits EnumButtonsComponent<TEnum, TContainer>

@if (Visible)
{
    <div class="btn-group-toggle" data-toggle="buttons">
        @foreach (var option in Options)
        {
            <label class="btn btn-primary @(Equals(Value, option.Value) ? "active" : "")">
                <input type="radio" name="@Name" value="@option.Value" checked="@(Equals(Value, option.Value))" onchange="@(() => Value = option.Value)"/>
                @option.Description
            </label>
        }
    </div>
}
```

And the class it's inheriting from:
```
public class EnumButtonsComponent<TEnum, TContainer> : EnumPickerComponentBase<TEnum, TContainer>
        where TEnum : struct
{
}

public class EnumPickerComponentBase<TEnum, TContainer> : AutoLabeledControlBase<TContainer>
    where TEnum : struct
{
    private string? stringValue;

    // without the constraint (or adding #nullable disable), TEnum is indeterminate. You cannot
    // mark it TEnum?, because it could be a value type. Not marking it as nullable raises a warning
    // because it could be null.
    private TEnum value = default;
    // implementation...
}
```

The ```EnumPickerComponentBase``` class needs the constraint due to #nullable limitations (have to pick either nullable or not nullable). Without this feature in Blazor, you would need to have an intermediate class for each type that you wanted an ```EnumPickerComponentBase```-derived template.

**Comment by springy76** [comment]
If Razor Components would create partial classes (last time I checked this was planned for 3.1.0) then you could define these generics including constraints in your own partial c# class.

**Comment by smartprogrammer93** [comment]
Isn't this solved now with 3.1 preview 1??? partial classes are published

**Comment by springy76** [comment]
Someone already tried this?

**Comment by stsrki** [comment]
I have just tried it and it doesn't work for generic components at all. It works for basic partial class.

### This works

```
public partial class CheckBox
{
    [Parameter] public RenderFragment ChildContent { get; set; }
}
```

### Doesn't work

```
public partial class CheckBox<TValue>
    where TValue : struct
{
    [Parameter] public TValue Checked { get; set; }

    [Parameter] public EventCallback<TValue> CheckedChanged { get; set; }

    [Parameter] public RenderFragment ChildContent { get; set; }
}
```

**Comment by smartprogrammer93** [comment]
This is a very important feature in my opinion. I wish if it can be done by .Net Core 3.1

**Comment by SQL-MisterMagoo** [comment]
@stsrki Can you explain what doesn't work?

If I add `@typeparam TValue` to a .razor file and add the same to the code behind partial class, with a constraint it works for me.

Razor file MyClass.razor
`@typeparam TValue`

Code behind MyClass.razor.cs
```
public partial class MyClass<TValue> where T:struct
```
Works as I would expect.

**Comment by rkone** [comment]
Thanks for the solution @SQL-MisterMagoo ! I didn't think about templates, as I just needed properties from the constraints for my control.

**Comment by miggat** [comment]
I used @SQL-MisterMagoo suggestion, and i have:
Test.razor:
`@typeparam TValue`

Test.razor.cs
`public partial class Test<TValue> where TValue: ISomeInterface`

But i get this error on build:
`Missing partial modifier on declaration of type 'Test'; another partial declaration of this type exists`

Also, notice the generated Test.razor.g.cs constructor
`public class Test<TValue> : Microsoft.AspNetCore.Components.ComponentBase`

I'm probably doing something wrong since i just landed into Blazor, but i can find the difference between your solution and mine

**Comment by FrancescoLorenzetti84** [comment]
I can confirm what @miggat said, I'm experiencing the same issue. This must be a bug.

**Comment by SQL-MisterMagoo** [comment]
@miggat you need to be on 3.1 for the partial classes to work

**Comment by FrancescoLorenzetti84** [comment]
@SQL-MisterMagoo I am on 3.1.0-preview4.19579.2 and I have that error

**Comment by SQL-MisterMagoo** [comment]
@miggat @FrancescoLorenzetti84 Here is a working sample for you

https://github.com/SQL-MisterMagoo/BlazorPartialConstraint

**Comment by miggat** [comment]
@FrancescoLorenzetti84 @SQL-MisterMagoo 
Apparently the problem was in visual studio itself. Upgrading to 16.4 solved the problem for me.

**Comment by xiety** [comment]
Type inference for component doesn't work when constraints are declared in a partial class.

Method CreateConstrainedComponent (in generated class TypeInference) is created without constraints so compiler produces error:

    CS0452 The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'ConstrainedComponent<T>'

Source from .razor.g.cs

    internal static class TypeInference
    {
        public static void CreateConstrainedComponent_0<T>(global::Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder __builder, int seq, int __seq0, T __arg0)
        {
        //error here
        __builder.OpenComponent<global::BlazorApp.Pages.ConstrainedComponent<T>>(seq);

**Comment by springy76** [comment]
At least (as a workaround) you can specifiy the type of T explicitly: `<ConstrainedComponent T="ConcreteTypeOfVariable" Value="@variable" .. />`

**Comment by joshlang** [comment]
> Type inference for component doesn't work when constraints are declared in a partial class.
> 
> Method CreateConstrainedComponent (in generated class TypeInference) is created without constraints so compiler produces error:
> 
> ```
> CS0452 The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'ConstrainedComponent<T>'
> ```
> 
> Source from .razor.g.cs
> 
> ```
> internal static class TypeInference
> {
>     public static void CreateConstrainedComponent_0<T>(global::Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder __builder, int seq, int __seq0, T __arg0)
>     {
>     //error here
>     __builder.OpenComponent<global::BlazorApp.Pages.ConstrainedComponent<T>>(seq);
> ```

^-- Indeed.  Just ran into this too

Tried to make a lil' component like this:
```
public partial class Fancy<TItem> where TItem : class
{
    [Parameter] public TItem? Item { get; set; }
    [Parameter] public RenderFragment<TItem> ChildContent { get; set; }
}

...(razor):

@ChildContent(Item)
````

Using it like this:

`<Fancy Item="..."> ... </Fancy>`

Results in `error CS0452: The type 'TItem' must be a reference type in order to use it as parameter 'TItem' in the generic type or method 'Fancy<TItem>'`

Because of auto-generated code like this:
```
    internal static class TypeInference
    {
        public static void CreateFancy_0<TItem>(global::Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder __builder, int seq, int __seq0, TItem __arg0, int __seq1, global::Microsoft.AspNetCore.Components.RenderFragment<TItem> __arg1)
...
```

^-- That needs to share the generic constraints, probably.

**Comment by Mike-EEE** [comment]
Starting to put my first Blazor application together, and while drafting out a simple CRUD Razor Component, quickly ran into this. 😭

![](https://i.imgur.com/1SYCDUc.png)

**Comment by EduVencovsky** [comment]
Not sure if this is the case, but I can't make a component because of this feature not being implemented. Also made a [SO question](https://stackoverflow.com/questions/60743109/how-to-make-a-generic-type-that-must-inherit-from-a-class-with-generic-type).

**Comment by musictopia2** [comment]
I have a question.  Would this also allow a case where if the constraint type is ComponentBase that it can create the ComponentBase in the razor syntax?  That would be needed because there can be times where a component list knows it needs to create a specific component based on a generic type and populate with the values needed.  Otherwise, a person is still forced to either create a special component for each type or the parent everytime has to create that component and populate the values even though the parent knows what is needed.

**Comment by mkArtakMSFT** [comment]
We've moved this to `Backlog` as this is also possible to achieve using partial classes.

**Comment by joshlang** [comment]
Unless something changed since Feb 25, no, it's not.

**Comment by poke** [comment]
@joshlang You may need to explicitly specify the type argument when you use your generic component. E.g.

```razor
<Fancy TItem="ItemType" Item="someItem"></Fancy>
```

**Comment by JayArrowz** [comment]
![image](https://user-images.githubusercontent.com/49910176/89132263-e107de80-d50a-11ea-8f2b-f09dea731754.png)
:(

**Comment by musictopia2** [comment]
Actually I found out I was able to achieve what was needed with partial classes.  In that case, maybe no need for changes because a person should use partial classes anyways.

**Comment by poke** [comment]
> In that case, maybe no need for changes because a person should use partial classes anyways.

I wouldn't agree with that since there are plenty examples of components that require a generic constraint but do not really need a code-behind.

It's good that we can achieve what we need right now, so this isn't a pressing issue. The end goal however should be that we can do this without code-behind. I'm totally fine if that comes later with 6.0.

**Comment by javiercn** [comment]
This has been implemented in a4a3668c455fe87f731803e4d0b3abfbc4f22198 and will be available in 6.0 preview4.
